test/Rings/GAlgebra-test.jl

```diff
@@ -0,0 +1,32 @@
+@testset "GAlgebra.printing" begin
+  r, (x, y, z) = QQ["x", "y", "z"]
+  R, (x, y, z) = g_algebra(r, [0 x*y x*z; 0 0 y*z + 1; 0 0 0], deglex(gens(r)))
```

Comment:
currently `@inferred` breaks here because the `S = elem_type(scr)` in `g_algebra`  is not inferred (even just for QQ).

Reply:
@thofma This related to `Singular.Rationals()`, which has abstract inferred type `Singular.Field`. How can we help @ code_warntype for
```
mutable struct Rationals <: Field
   ptr::libSingular.coeffs_ptr
   refcount::Int

   function Rationals()
      return get!(RationalsID, :QQ) do
         ptr = libSingular.nInitChar(libSingular.n_Q, Ptr{Nothing}(0))
         d = new(ptr, 0)
         RationalsID[:QQ] = d
         finalizer(_Ring_finalizer, d)
         return d
      end
   end
end
```

Reply:
To answer this I need to ask some other questions. Please look at the chat if possible.

Reply:
Or why not put the dict after the struct and use `Dict{Symbol, Rationals}`?

Reply:
ok, but why doesn't
````
function Rationals()::Rationals
   ....
end
````
do the trick?

Reply:
Don't know. Does `@code_lowered` show some `typeassert` with `Rationals` at the end in this case?

Reply:
@code lowered shows no type assert. Is this a bug?
```
struct foo
  a::Int
  function foo()::foo
    return 2
  end
end

typeof(foo())
```

Reply:
It seems to be fixed in julia master, probably fixed in 1.8.